onnxruntime/core/optimizer/constant_folding.cc

```diff
@@ -12,104 +13,158 @@ using namespace onnxruntime::common;
-    if (!graph_utils::IsSupportedProvider(*node, GetCompatibleExecutionProviders()) ||
-        excluded_op_types_.find(node->OpType()) != excluded_op_types_.end() ||
-        // constant folding does not support executing a node that includes subgraphs (control flow operators,
-        // such as If/Loop/Scan, fall into this category). individual nodes in the subgraph will be processed
-        // by the Recurse call above
-        node->ContainsSubgraph() ||
-        !graph_utils::AllNodeInputsAreConstant(graph, *node, constant_inputs)) {
-      continue;
-    }
-
-    // override the EP while setting up OptimizerExecutionFrame::Info so that it will use the CPU kernel for Compute.
-    if (!cpu_ep) {
-      node->SetExecutionProviderType(kCpuExecutionProvider);
-    }
-
-    // Create execution frame for executing constant nodes.
-    OptimizerExecutionFrame::Info info({node}, constant_inputs);
-
-    // undo the EP change in case something fails prior to node removal
-    if (!cpu_ep) {
-      node->SetExecutionProviderType(ep_type);
-    }
-
-    std::vector<int> fetch_mlvalue_idxs;
-    for (const auto* node_out : node->OutputDefs()) {
-      fetch_mlvalue_idxs.push_back(info.GetMLValueIndex(node_out->Name()));
-    }
+    // Resolve graph to force run shape and type inference.
+    ORT_ENFORCE(graph.Resolve().IsOK());
 
-    OptimizerExecutionFrame frame(info, fetch_mlvalue_idxs);
+    GraphViewer graph_viewer(graph);
+    auto& order = graph_viewer.GetNodesInTopologicalOrder();
 
-    auto* kernel = info.GetKernel(node->Index());
-    OpKernelContext op_kernel_context(&frame, kernel, nullptr, onnxruntime::logging::LoggingManager::DefaultLogger());
+    for (NodeIndex i : order) {
+      auto* node = graph.GetNode(i);
+      if (!node) {
+        continue;
+      }
 
-    ORT_RETURN_IF_ERROR(kernel->Compute(&op_kernel_context));
+      ORT_RETURN_IF_ERROR(Recurse(*node, modified, graph_level, logger));
+
+      bool convert_to_constant = false;
+      if (node->OpType().compare("Shape") == 0) {
+        auto shape = node->MutableInputDefs()[0]->Shape();
+        bool is_concrete_shape = true;
+        std::vector<int64_t> dim_values;
+        if (shape != nullptr && shape->dim_size() > 0) {
+          for (int i = 0; i < shape->dim_size(); i++) {
+            auto dim = shape->dim(i);
+            if (!utils::HasDimValue(dim) || dim.dim_value() < 1) {
+              is_concrete_shape = false;
+              break;
+            }
+            dim_values.push_back(dim.dim_value());
+          }
+        } else {
+          is_concrete_shape = false;
+        }
+
+        if (is_concrete_shape) {
+          ONNX_NAMESPACE::TensorProto shape_constant;
+          auto* constant_arg_out = node->MutableOutputDefs()[0];
+          shape_constant.set_name(constant_arg_out->Name());
+          shape_constant.set_data_type(ONNX_NAMESPACE::TensorProto_DataType_INT64);
+          shape_constant.add_dims(dim_values.size());
```

Comment:
>dim_values [](start = 34, length = 10)

will it work with scalar that dim_values.size() == 0?

Reply:
Add support for this case.